### PR TITLE
refactor: Cosmos indexers

### PR DESCRIPTION
**Pull request by ameten**
### Description

Refactoring Cosmos related indexers:
* Separate dispatch and deliver indexers
* Re-use repeated code on log parsing

### Backward compatibility

Yes

### Testing

E2E Test for Cosmos

### CODE REVIEW

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+/// A reference to a Mailbox contract on some Cosmos chain
+pub struct CosmosMailbox {
+    config: ConnectionConf,
+    domain: HyperlaneDomain,
+    address: H256,
+    provider: CosmosProvider,
+}
+
+impl CosmosMailbox {
+    /// Create a reference to a mailbox at a specific Ethereum address on some
```
Comment (tkporter) [inline rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:35]: nit: I know this was already like this but a drive-by fix
```suggestion
    /// Create a reference to a mailbox at a specific Cosmos address on some
```

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl Debug for CosmosMailbox {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        todo!()
```
Comment (tkporter) [inline rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:85]: lol - also realize this already existed but this feels dangerous, could you fix this as part of your efforts? doesn't need to be in this PR

Could definitely see us accidentally logging the Mailbox and getting a panic 😅😭

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            .grpc()
+            .wasm_query(GeneralMailboxQuery { mailbox: payload }, None)
+            .await
+        {
+            Ok(v) => {
+                let response: payloads::mailbox::DeliveredResponse = serde_json::from_slice(&v)?;
+
+                response.delivered
+            }
+            Err(err) => {
```
Comment (tkporter) [inline rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:115]: as I'm sure you've noticed the cosmos implementation is very ugly, so definitely feel free to make tweaks along the way (as you're already doing 😅). The impl was largely done outside our team and we did our best to clean it up from a functionality POV but less from a code elegance / maintenance POV. But reading this now I can't imagine why we have this match at all instead of just propagating the error

File: rust/chains/hyperlane-cosmos/src/merkle_tree_hook.rs [hunk 4]
```diff
@@ -283,22 +284,12 @@ impl Indexer<MerkleTreeInsertion> for CosmosMerkleTreeHookIndexer {
...
-                            block_number,
-                            Self::merkle_tree_insertion_parser,
-                            "MerkleTreeInsertionCursor",
-                        )
-                        .await;
-                    (logs, block_number)
-                })
-            })
-            .collect();
+        let logs_futures = parse_logs_in_range(
```
Comment (tkporter) [inline rust/chains/hyperlane-cosmos/src/merkle_tree_hook.rs:287]: nit: happy either way, but could even take it further and just have all these `fetch_logs_in_range` fns call a single function that also does `execute_and_parse_log_futures` for you too

**Review by tkporter — APPROVED** [review]
loving the cleanup, ty 🙏

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs (hunk 1 above)
Comment (ameten) [reply rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:35]: Done

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs (hunk 3 above)
Comment (ameten) [reply rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:115]: Fixed.

File: rust/chains/hyperlane-cosmos/src/mailbox/contract.rs (hunk 2 above)
Comment (ameten) [reply rust/chains/hyperlane-cosmos/src/mailbox/contract.rs:85]: Fixed.

File: rust/chains/hyperlane-cosmos/src/merkle_tree_hook.rs (hunk 4 above)
Comment (ameten) [reply rust/chains/hyperlane-cosmos/src/merkle_tree_hook.rs:287]: I'll keep it as is for the moment.
